<img src="https://s8.hostingkartinok.com/uploads/images/2018/08/308b49fcfbc619d629fe4604bceb67ac.jpg" width=500,>
<h3 style="text-align: center;"><b>Физтех-Школа Прикладной математики и информатики (ФПМИ) МФТИ</b></h3>

---

<h2 style="text-align: center;"><b>PyTorch. Основы: синтаксис, torch.cuda и torch.autograd</b></h2>

<p style="align: center;"><img src="https://raw.githubusercontent.com/pytorch/pytorch/master/docs/source/_static/img/pytorch-logo-dark.png", width=400, height=300></p>

На этом занятии мы рассмотрим основы фреймворка глубокого обучения PyTorch.  

Когда хочется написать какую-нибудь нейросеть, решающую определённую задачу, будь то какая-нибудь простая классификация чего-либо или обнаружение лиц людей на видео. Всё, конечно, всегда начинается со **сбора данных**, а уже потом реализуются модели и проводятся эксперименты.  

Однако люди быстро поняли, что писать свои нейронные сети каждый раз с нуля ну очень уж долго и неразумно, поэтому придумали так называемые **фреймворки** - модули, в которых есть функционал, с помощью которого можно быстро и просто решать типовые задачи, и уже с помощью этих средств писать решения к более сложным задачам.

Есть много различных фремворков глубокого обучения. Разница между ними прежде всего в том, каков общий принцип вычислений. 
Например, в **Caffe и Caffe2** вы пишете код, по сути, составляя его из готовых "кусочков", как в Lego, в **TensorFlow и Theano** вы сначала объявляете вычислительный граф, потом компилируете его и запускаете (sees.run()), в то время как в **Torch и PyTorch** вы пишете почти точно так же, как на NumPy, а граф вычислений создаётся только при запуске (то есть существует только во время выполнения, потом он "разрушается"). **Keras** позволяет как строить блоки, так и компилировать свой граф:

<p style="align: center;"><img src="https://habrastorage.org/web/e3e/c3e/b78/e3ec3eb78d714a7993a6b922911c0866.png", width=500, height=500></p>  
<p style="text-align: center;"><i>Картинка взята из отличной [статьи на Хабре](https://habr.com/post/334380/)</i><p>

<h3 style="text-align: center;"><b>Установка</b></h3>

Инструкция по установке PyTorch есть на официальном [Github'e Deep Learning School](https://github.com/deepmipt/dlschl/wiki/%D0%98%D0%BD%D1%81%D1%82%D1%80%D1%83%D0%BA%D1%86%D0%B8%D1%8F-%D0%BF%D0%BE-%D1%83%D1%81%D1%82%D0%B0%D0%BD%D0%BE%D0%B2%D0%BA%D0%B5-PyTorch).

<h3 style="text-align: center;">Синтаксис<b></b></h3>

In [2]:
import torch

Сначала немного фактов про PyTorch:  
- динамический граф вычислений
- удобные модули `torch.nn` и `torchvision для` написания нейросеток с минимальными усилиями
- в некоторых задачах даже быстрее TensorFlow (но не во всех)  
- легко проводить вычисления на GPU

Если PyTorch представить формулой, то она будет такой:  

$$PyTorch = NumPy + CUDA + Autograd$$

(CUDA - [wiki](https://ru.wikipedia.org/wiki/CUDA))

Посмотрим, как в PyTorch выполняются операции с векторами.  

Напоминание: **тензором** называется многомерный вектор, то есть есть:  

x = np.array([1,2,3]) - вектор = тензор размерности 1 (то есть (1,))  
y = np.array([[1, 2, 3], [4, 5, 6]]) - матрица = тензор размерности 2 (в данном случае тензор (2, 3))  
z = np.array([[[1, 2, 3], [4, 5, 6], [7, 8, 9]],  
    [[1, 2, 3], [4, 5, 6], [7, 8, 9]],  
              [[1, 2, 3], [4, 5, 6], [7, 8, 9]]]) - "кубик" (3, 3, 3) = тензор размерности 3 (в данном случае (3, 3, 3))

Простейшим примером 3-мерного тензора является **картинка** - это "параллелепипед" из чисел, у коготорого три размерности - высота, ширина и количество каналов, значит это тензор размерности 3.

Понятие тензора нужно знать потому, что в PyTorch мы оперируем переменными типа `torch.Tensor` (`FloatTensor`, `IntTensor`, `ByteTensor`), и пугаться их названий совершенно не нужно - это просто векторы, у которых несколько размерностей.

Все типы тензоров:

In [6]:
torch.HalfTensor      # 16 бит, с плавающей точкой
torch.FloatTensor     # 32 бита,  с плавающей точкой
torch.DoubleTensor    # 64 бита, с плавающей точкой

torch.ShortTensor     # 16 бит, целочисленный, знаковый
torch.IntTensor       # 32 бита, целочисленный, знаковый
torch.LongTensor      # 64 бита, целочисленный, знаковый

torch.CharTensor      # 8 бит, целочисленный, знаковый
torch.ByteTensor      # 8 бит, целочисленный, беззнаковый

torch.ByteTensor

Мы будем использовать только `torch.FloatTensor()` и `torch.IntTensor()`. 

Перейдём к делу:

* Создание тензоров:

In [7]:
a = torch.FloatTensor([1, 2])
a

tensor([1., 2.])

In [8]:
a.shape

torch.Size([2])

In [9]:
b = torch.FloatTensor([[1,2,3], [4,5,6]])
b

tensor([[1., 2., 3.],
        [4., 5., 6.]])

In [10]:
b.shape

torch.Size([2, 3])

In [11]:
x = torch.FloatTensor(2,3,4)

In [12]:
x

tensor([[[-4.1822e-02,  9.3747e-43,  8.9683e-44,  0.0000e+00],
         [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
         [ 0.0000e+00,  0.0000e+00,  2.0775e+20,  2.6806e-09]],

        [[ 2.1972e-04,  4.2658e-08,  1.3000e+22,  1.3001e+22],
         [ 2.5964e+20,  4.2466e+21,  1.0548e-08,  1.7585e-04],
         [ 1.1039e-05,  5.2002e+22,  1.0733e-08,  8.3017e-10]]])

In [13]:
x = torch.FloatTensor(100)
x

tensor([        nan,         nan,  2.8026e-45,  0.0000e+00,  1.4910e-42,
         0.0000e+00,  5.6052e-45,  0.0000e+00,  1.5414e-42,  0.0000e+00,
         5.6052e-45,  0.0000e+00,  1.1659e-42,  0.0000e+00,  5.7593e-43,
         0.0000e+00,  1.0930e-42,  0.0000e+00,  5.7593e-43,  0.0000e+00,
         1.2331e-42,  0.0000e+00,  5.7593e-43,  0.0000e+00,  1.1547e-42,
         0.0000e+00,  5.7593e-43,  9.3747e-43,  1.4854e-42,  0.0000e+00,
         5.7593e-43,  0.0000e+00,  1.4910e-42,  0.0000e+00,  5.7593e-43,
         0.0000e+00,  1.4742e-42,  0.0000e+00,  5.7593e-43,  0.0000e+00,
         1.5414e-42,  0.0000e+00,  5.7593e-43,  0.0000e+00,  1.5582e-42,
         0.0000e+00,  5.7593e-43,  0.0000e+00,  1.4910e-42,  0.0000e+00,
         5.6052e-45,  0.0000e+00,  1.5414e-42,  0.0000e+00,  5.6052e-45,
         0.0000e+00, -1.0846e+01,  4.5911e-41, -1.0846e+01,  4.5911e-41,
         0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,
         0.0000e+00,  0.0000e+00,  0.0000e+00, -3.5

In [14]:
x = torch.IntTensor(45, 57, 14, 2)
x.shape

torch.Size([45, 57, 14, 2])

**Обратите внимание** - если вы создаёте тензор через задание размерностей (как в примере выше), то он изначально заполняюстя случайным "мусором". Что инициализировать нулями, нужно написать .zero_() в конце:

In [15]:
x = torch.IntTensor(3, 2, 4)
x

tensor([[[-1121235416,         669,          64,           0],
         [          0,           0,           0,           0]],

        [[          0,           0,   929116771,   942683445],
         [  909141301,   842412130,  1647719781,   892953145]],

        [[  842217265,   963000629,  1681482295,  1631150393],
         [ 1650668852,   912406322,  1664299312,   859322468]]],
       dtype=torch.int32)

In [16]:
x = torch.IntTensor(3, 2, 4).zero_()
x

tensor([[[0, 0, 0, 0],
         [0, 0, 0, 0]],

        [[0, 0, 0, 0],
         [0, 0, 0, 0]],

        [[0, 0, 0, 0],
         [0, 0, 0, 0]]], dtype=torch.int32)

Аналог функции `np.reshape()` == `torch.view()`:

In [17]:
b.view(3, 2)

tensor([[1., 2.],
        [3., 4.],
        [5., 6.]])

In [18]:
b

tensor([[1., 2., 3.],
        [4., 5., 6.]])

**Обратите внимание** - torch.view() создаёт новый тензор, а не изменяет старый!

In [19]:
b.view(-1)

tensor([1., 2., 3., 4., 5., 6.])

In [20]:
b

tensor([[1., 2., 3.],
        [4., 5., 6.]])

* Изменение типа тензора:

In [21]:
a = torch.FloatTensor([1.5, 3.2, -7])

In [22]:
a.type_as(torch.IntTensor())

tensor([ 1,  3, -7], dtype=torch.int32)

In [23]:
a.type_as(torch.ByteTensor())

tensor([  1,   3, 249], dtype=torch.uint8)

Обратите внимание, что при `.type_as()` создаётся новый тензор (старый не меняется), то есть это не in-place операция:

In [24]:
a

tensor([ 1.5000,  3.2000, -7.0000])

* Индексация точная такая же, как и в NumPy:

In [25]:
a = torch.FloatTensor([[100, 20, 35], [15, 163, 534], [52, 90, 66]])
a

tensor([[100.,  20.,  35.],
        [ 15., 163., 534.],
        [ 52.,  90.,  66.]])

In [26]:
a[0, 0]

tensor(100.)

In [27]:
a[0][0]

tensor(100.)

In [28]:
a[0:2, 0:2]

tensor([[100.,  20.],
        [ 15., 163.]])

### Задача 1

1). Создайте два вещественных тензора: `a` размером (3, 4) и `b` размером (12,)   
2). Создайте тензор `c`, являющийся тензором `b`, но размера (2, 2, 3)  
3). Выведите первый столбец матрицы `a` с помощью индексации

In [29]:
a = torch.FloatTensor(3, 4)
b = torch.FloatTensor(range(12))

In [30]:
print(a)
print(b)

tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])
tensor([ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11.])


In [31]:
c = b.view(2, 2, 3)
c

tensor([[[ 0.,  1.,  2.],
         [ 3.,  4.,  5.]],

        [[ 6.,  7.,  8.],
         [ 9., 10., 11.]]])

In [32]:
a[:, 0]

tensor([0., 0., 0.])

**Арифметика и булевы операции** работаю также, как и в NumPy, **НО** лучше использовать не опреаторы `+`, `-`, `*`, `/`, а их аналоги:  

| Оператор | Аналог |
|:-:|:-:|
|`+`| `torch.add()` |
|`-`| `torch.sub()` |
|`*`| `torch.mul()` |
|`/`| `torch.div()` |

* Сложение:

In [33]:
a = torch.FloatTensor([[1, 2, 3], [10, 20, 30], [100, 200, 300]])
b = torch.FloatTensor([[-1, -2, -3], [-10, -20, -30], [100, 200, 300]])

In [34]:
a + b

tensor([[  0.,   0.,   0.],
        [  0.,   0.,   0.],
        [200., 400., 600.]])

Лучше:

In [35]:
a.add(b)

tensor([[  0.,   0.,   0.],
        [  0.,   0.,   0.],
        [200., 400., 600.]])

In [36]:
b = -a
b

tensor([[  -1.,   -2.,   -3.],
        [ -10.,  -20.,  -30.],
        [-100., -200., -300.]])

In [37]:
a + b

tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])

* Вычитание:

In [38]:
a - b

tensor([[  2.,   4.,   6.],
        [ 20.,  40.,  60.],
        [200., 400., 600.]])

Лучше:

In [39]:
a.sub(b)

tensor([[  2.,   4.,   6.],
        [ 20.,  40.,  60.],
        [200., 400., 600.]])

* Умножение (поэлементное):

In [40]:
a * b

tensor([[-1.0000e+00, -4.0000e+00, -9.0000e+00],
        [-1.0000e+02, -4.0000e+02, -9.0000e+02],
        [-1.0000e+04, -4.0000e+04, -9.0000e+04]])

Лучше:

In [41]:
a.mul(b)

tensor([[-1.0000e+00, -4.0000e+00, -9.0000e+00],
        [-1.0000e+02, -4.0000e+02, -9.0000e+02],
        [-1.0000e+04, -4.0000e+04, -9.0000e+04]])

* Деление (поэлементное):

In [42]:
a = torch.FloatTensor([[1, 2, 3], [10, 20, 30], [100, 200, 300]])
b = torch.FloatTensor([[-1, -2, -3], [-10, -20, -30], [100, 200, 300]])

In [43]:
a / b

tensor([[-1., -1., -1.],
        [-1., -1., -1.],
        [ 1.,  1.,  1.]])

Лучше:

In [44]:
a.div(b)

tensor([[-1., -1., -1.],
        [-1., -1., -1.],
        [ 1.,  1.,  1.]])

Заметьте, все эти операции **не меняют исходные тензоры**, а **создают новые**:

In [45]:
a

tensor([[  1.,   2.,   3.],
        [ 10.,  20.,  30.],
        [100., 200., 300.]])

In [46]:
b

tensor([[ -1.,  -2.,  -3.],
        [-10., -20., -30.],
        [100., 200., 300.]])

### Задача 2

1). Создайте два вещественных тензора: `a` размером (5, 2) и `b` размером (1,10)   
2). Создайте тензор `c`, являющийся тензором `b`, но размера (5, 2)  
3). Произведите все арифметические операции с тензорами `a` и `c`

In [47]:
a = torch.FloatTensor(5, 2)
b = torch.FloatTensor(1, 10)
b

tensor([[-3.2920e-04,  9.3747e-43, -9.4688e+16,  5.7453e-44, -1.0010e-03,
          9.3747e-43,  2.9287e-43,  0.0000e+00,  0.0000e+00,  0.0000e+00]])

In [48]:
c = b.view(5, 2)
c

tensor([[-3.2920e-04,  9.3747e-43],
        [-9.4688e+16,  5.7453e-44],
        [-1.0010e-03,  9.3747e-43],
        [ 2.9287e-43,  0.0000e+00],
        [ 0.0000e+00,  0.0000e+00]])

In [49]:
sum = a.add(c)
dif = a.sub(c)
mul = a.mul(c)
div = a.div(c)
print(sum)
print(dif)
print(mul)
print(div)

tensor([[-6.5839e-04,  1.8749e-42],
        [-9.4688e+16,  1.0229e-43],
        [-1.0010e-03,  9.6836e+24],
        [ 1.7748e+28,  0.0000e+00],
        [ 0.0000e+00,  0.0000e+00]])
tensor([[ 0.0000e+00,  0.0000e+00],
        [ 9.4688e+16, -1.2612e-44],
        [ 1.0010e-03,  9.6836e+24],
        [ 1.7748e+28,  0.0000e+00],
        [ 0.0000e+00,  0.0000e+00]])
tensor([[ 1.0837e-07,  0.0000e+00],
        [-5.7055e-27,  0.0000e+00],
        [-0.0000e+00,  9.0781e-18],
        [ 5.1980e-15,  0.0000e+00],
        [ 0.0000e+00,  0.0000e+00]])
tensor([[ 1.0000e+00,  1.0000e+00],
        [-0.0000e+00,  7.8049e-01],
        [-1.3719e-40,         inf],
        [        inf,         nan],
        [        nan,         nan]])


* **Операторы сравнения**:

In [50]:
a = torch.FloatTensor([[1, 2, 3], [10, 20, 30], [100, 200, 300]])
b = torch.FloatTensor([[-1, -2, -3], [-10, -20, -30], [100, 200, 300]])

In [51]:
a == b

tensor([[0, 0, 0],
        [0, 0, 0],
        [1, 1, 1]], dtype=torch.uint8)

In [52]:
a != b

tensor([[1, 1, 1],
        [1, 1, 1],
        [0, 0, 0]], dtype=torch.uint8)

In [53]:
a < b

tensor([[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]], dtype=torch.uint8)

In [54]:
a > b

tensor([[1, 1, 1],
        [1, 1, 1],
        [0, 0, 0]], dtype=torch.uint8)

* **Булевы маски**:

In [55]:
a[a > b]

tensor([ 1.,  2.,  3., 10., 20., 30.])

In [56]:
b[a == b]

tensor([100., 200., 300.])

Опять же, тензоры не меняются:

In [57]:
a

tensor([[  1.,   2.,   3.],
        [ 10.,  20.,  30.],
        [100., 200., 300.]])

In [58]:
b

tensor([[ -1.,  -2.,  -3.],
        [-10., -20., -30.],
        [100., 200., 300.]])

Применение **стандартных функций** такое же, как и в numpy - поэлементное:

In [59]:
a = torch.FloatTensor([[1, 2, 3], [10, 20, 30], [100, 200, 300]])

In [60]:
a.sin()

tensor([[ 0.8415,  0.9093,  0.1411],
        [-0.5440,  0.9129, -0.9880],
        [-0.5064, -0.8733, -0.9998]])

In [61]:
torch.sin(a)

tensor([[ 0.8415,  0.9093,  0.1411],
        [-0.5440,  0.9129, -0.9880],
        [-0.5064, -0.8733, -0.9998]])

In [62]:
a.cos()

tensor([[ 0.5403, -0.4161, -0.9900],
        [-0.8391,  0.4081,  0.1543],
        [ 0.8623,  0.4872, -0.0221]])

In [63]:
a.exp()

tensor([[2.7183e+00, 7.3891e+00, 2.0086e+01],
        [2.2026e+04, 4.8517e+08, 1.0686e+13],
        [       inf,        inf,        inf]])

In [64]:
a.log()

tensor([[0.0000, 0.6931, 1.0986],
        [2.3026, 2.9957, 3.4012],
        [4.6052, 5.2983, 5.7038]])

In [65]:
b = -a
b

tensor([[  -1.,   -2.,   -3.],
        [ -10.,  -20.,  -30.],
        [-100., -200., -300.]])

In [66]:
b.abs()

tensor([[  1.,   2.,   3.],
        [ 10.,  20.,  30.],
        [100., 200., 300.]])

**Сумма, среднее, максимум, минимум**:

In [67]:
a.sum()

tensor(666.)

In [68]:
a.mean()

tensor(74.)

По осям:

In [69]:
a.sum(0)

tensor([111., 222., 333.])

In [70]:
a.sum(1)

tensor([  6.,  60., 600.])

In [71]:
a.max()

tensor(300.)

In [72]:
a.max(0)

(tensor([100., 200., 300.]), tensor([2, 2, 2]))

In [73]:
a.min()

tensor(1.)

In [74]:
a.min(0)

(tensor([1., 2., 3.]), tensor([0, 0, 0]))

Обратите внимание - второй тензор при вызове функций .max() и .min() - это индексы этих максимальных/минимальных элементов по указанной размерности (то есть в данном случае a.min() вернул (1, 2, 3) - минимумы по 0 оси (по столбцам), и их индексы по 0-ой оси (0,0,0) (номер каждого элемента в своём столбце)).

### Задача 3

Создайте тензор `a` размерности (100, 780, 780, 3) (можно интерпретировать это как 100 картинок размера 780х780 с тремя цветовыми каналами) и выведите первый элемент этого тензора как картинку (с помощью matplotlib.pyplot).

Выведите среднее элементов по 1-ой оси (по сути - средняя картинка по всем картинкам) и по 4-ой оси (по сути - усреднение каналов для каждой картинки).

In [75]:
a = torch.FloatTensor(100, 780, 780, 3)
a.mean(0).shape

torch.Size([780, 780, 3])

**Матричные операции:**

In [ ]:
z = x.mm(y)
z = torch.mm(x, y)
# Матричное умножение.
z = x.mv(v)
z = torch.mv(x, v)
# Умножение матрицы на вектор.
z = x.dot(y)
z = torch.dot(x, y)
# Скалярное умножение тензоров.
bz = bx.bmm(by)
bz = torch.bmm(bx, by)
# Перемножает матрицы целыми батчами.

* Транспонирование матрицы (тензора):

In [77]:
a = torch.FloatTensor([[1, 2, 3], [10, 20, 30], [100, 200, 300]])
a

tensor([[  1.,   2.,   3.],
        [ 10.,  20.,  30.],
        [100., 200., 300.]])

In [78]:
a.t()

tensor([[  1.,  10., 100.],
        [  2.,  20., 200.],
        [  3.,  30., 300.]])

И снова - сам тензор не меняется (то есть при вызове создаётся новый):

In [79]:
a

tensor([[  1.,   2.,   3.],
        [ 10.,  20.,  30.],
        [100., 200., 300.]])

* Скалярное произведение векторов (1-мерных тензоров):

In [80]:
a = torch.FloatTensor([1, 2, 3, 4, 5, 6])
b = torch.FloatTensor([-1, -2, -4, -6, -8, -10])

In [81]:
a.dot(b)

tensor(-141.)

In [82]:
a @ b

tensor(-141.)

In [83]:
type(a)

torch.Tensor

In [84]:
type(b)

torch.Tensor

In [85]:
type(a @ b)

torch.Tensor

* Матричное умножение:

In [86]:
a = torch.FloatTensor([[1, 2, 3], [10, 20, 30], [100, 200, 300]])
b = torch.FloatTensor([[-1, -2, -3], [-10, -20, -30], [100, 200, 300]])

In [87]:
a.mm(b)

tensor([[  279.,   558.,   837.],
        [ 2790.,  5580.,  8370.],
        [27900., 55800., 83700.]])

In [88]:
a @ b

tensor([[  279.,   558.,   837.],
        [ 2790.,  5580.,  8370.],
        [27900., 55800., 83700.]])

Тензоры неизменны:

In [89]:
a

tensor([[  1.,   2.,   3.],
        [ 10.,  20.,  30.],
        [100., 200., 300.]])

In [90]:
b

tensor([[ -1.,  -2.,  -3.],
        [-10., -20., -30.],
        [100., 200., 300.]])

In [91]:
a = torch.FloatTensor([[1, 2, 3], [10, 20, 30], [100, 200, 300]])
b = torch.FloatTensor([[-1], [-10], [100]])

In [92]:
print(a.shape, b.shape)

torch.Size([3, 3]) torch.Size([3, 1])


In [93]:
a @ b

tensor([[  279.],
        [ 2790.],
        [27900.]])

Если "развернуть" тензор b просто в массив элементов (`torch.view(-1)`), умножение будет как на столбец:

In [94]:
b

tensor([[ -1.],
        [-10.],
        [100.]])

In [95]:
b.view(-1)

tensor([ -1., -10., 100.])

In [96]:
a @ b.view(-1)

tensor([  279.,  2790., 27900.])

In [97]:
a.mv(b.view(-1))

tensor([  279.,  2790., 27900.])

**Перевод из NumPy в PyTorch**:

In [98]:
import numpy as np

a = np.random.rand(3, 3)
a

array([[0.87122966, 0.77548287, 0.99435012],
       [0.9561005 , 0.83148683, 0.35000053],
       [0.5885051 , 0.05767388, 0.47980665]])

In [99]:
b = torch.from_numpy(a)
b

tensor([[0.8712, 0.7755, 0.9944],
        [0.9561, 0.8315, 0.3500],
        [0.5885, 0.0577, 0.4798]], dtype=torch.float64)

**НО!** Обратите внимание - a и b в этом случае будут использовать одно и то же хранилище данных, то есть измение одного тензора будет менять и другой:

In [100]:
b -= b
b

tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]], dtype=torch.float64)

In [101]:
a

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

**Перевод из PyTorch в NumPy:**

In [102]:
a = torch.FloatTensor(2, 3, 4)
a

tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])

In [103]:
type(a)

torch.Tensor

In [104]:
x = a.numpy()
x

array([[[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]],

       [[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]]], dtype=float32)

In [105]:
x.shape

(2, 3, 4)

In [106]:
type(x)

numpy.ndarray

### Задача 4

Напишите функцию `forward_pass(X, w)` ($w_0$ входит в $w$) для одного нейрона (с сигмоидой) с помощью PyTorch (у вас уже должен быть код на NumPy).

In [107]:
def sigmoid(x):
    """Сигмоидальная функция"""
    return 1 / (1 + np.exp(-x))

In [108]:
def forward_pass(X, w):
    n = X.shape[0]
    y_pred = torch.FloatTensor(n, 1)
    y_pred = sigmoid(X.mm(w).view(-1,1))
    return(y_pred)

In [109]:
X = torch.FloatTensor([[-5, 5], [2, 3], [1, -1]])
w = torch.FloatTensor([[-0.5], [2.5]])
result = forward_pass(X, w)
print('result: {}'.format(result))

result: tensor([[1.0000],
        [0.9985],
        [0.0474]])


Должно получиться: 

|variable|value|
|-|-|
|**X**|torch.FloatTensor([[-5, 5], [15, 20], [100, -700]])|
|**w**|torch.FloatTensor([[-0.5], [150]])|
|**result**|torch.FloatTensor([[1.0000], [0.9985], [0.0474]])|   

<h3 style="text-align: center;">[CUDA](https://ru.wikipedia.org/wiki/CUDA)<b></b></h3>

[Краткое видео про то, как GPU используется в обучении нейросетей](https://www.youtube.com/watch?v=EobhK0UZm80)

Все вычисления в PyTorch можно проводить как на CPU, так и на GPU (Graphical Processing Unit) (если она у вас есть). В PyTorch переключение между ними делается очень просто, что является одной из ключевых его особенностей.

In [110]:
x = torch.FloatTensor(1024, 1024).uniform_()
x

tensor([[0.4230, 0.4093, 0.6542,  ..., 0.7091, 0.5403, 0.7385],
        [0.8526, 0.0909, 0.3441,  ..., 0.1869, 0.2340, 0.0576],
        [0.2086, 0.0607, 0.9921,  ..., 0.8893, 0.9689, 0.1569],
        ...,
        [0.7695, 0.2648, 0.7162,  ..., 0.7105, 0.8771, 0.6484],
        [0.0426, 0.7683, 0.1250,  ..., 0.9431, 0.1050, 0.1511],
        [0.1540, 0.3111, 0.7384,  ..., 0.2454, 0.6582, 0.5785]])

In [111]:
x.is_cuda

False

Переместим на GPU:

In [112]:
x = x.cuda()

In [115]:
x.is_cuda

True

In [116]:
x

tensor([[0.4230, 0.4093, 0.6542,  ..., 0.7091, 0.5403, 0.7385],
        [0.8526, 0.0909, 0.3441,  ..., 0.1869, 0.2340, 0.0576],
        [0.2086, 0.0607, 0.9921,  ..., 0.8893, 0.9689, 0.1569],
        ...,
        [0.7695, 0.2648, 0.7162,  ..., 0.7105, 0.8771, 0.6484],
        [0.0426, 0.7683, 0.1250,  ..., 0.9431, 0.1050, 0.1511],
        [0.1540, 0.3111, 0.7384,  ..., 0.2454, 0.6582, 0.5785]],
       device='cuda:0')

Перемножим две тензора на GPu и вернём результат вычисления на CPU:

In [117]:
a = torch.FloatTensor(10000, 10000).uniform_()
b = torch.FloatTensor(10000, 10000).uniform_()
c = a.cuda().mul(b.cuda()).cpu()

In [118]:
c

tensor([[0.5148, 0.4447, 0.1388,  ..., 0.4346, 0.0377, 0.2424],
        [0.2516, 0.1703, 0.0247,  ..., 0.0187, 0.4402, 0.0126],
        [0.5699, 0.5254, 0.2331,  ..., 0.2444, 0.1085, 0.0165],
        ...,
        [0.0143, 0.1268, 0.2105,  ..., 0.1687, 0.1774, 0.0398],
        [0.5336, 0.6748, 0.2809,  ..., 0.1018, 0.2622, 0.7407],
        [0.0549, 0.1738, 0.3901,  ..., 0.4950, 0.4861, 0.4821]])

In [119]:
a

tensor([[0.8609, 0.5007, 0.7149,  ..., 0.8162, 0.4706, 0.8043],
        [0.3785, 0.3317, 0.0312,  ..., 0.5537, 0.6043, 0.0300],
        [0.7362, 0.9857, 0.3071,  ..., 0.9041, 0.2431, 0.0849],
        ...,
        [0.1508, 0.5319, 0.7457,  ..., 0.8752, 0.5593, 0.4286],
        [0.7425, 0.6781, 0.3367,  ..., 0.3008, 0.7062, 0.8175],
        [0.2489, 0.3515, 0.7246,  ..., 0.9771, 0.5912, 0.5755]])

Тензоры, лежащие на CPU, и тензоры, лежащие на GPU, недоступны друг для друга:

In [120]:
a = torch.FloatTensor(1000, 1000).uniform_().cpu()
b = torch.FloatTensor(10000, 10000).uniform_().cuda()

In [122]:
a + b

RuntimeError: The size of tensor a (1000) must match the size of tensor b (10000) at non-singleton dimension 1

Вот ещё немного про то, как можно работать с GPU:

In [123]:
x = torch.FloatTensor(5, 5, 5).uniform_()

# проверяем, есть ли CUDA (то есть NVidia GPU)
if torch.cuda.is_available():
    # так можно получить имя устройства, которое связано с CUDA
    # (полезно в случае с несколькими видеокартами)
    device = torch.device('cuda')          # CUDA-device объект
    y = torch.ones_like(x, device=device)  # создаём тензор на GPU
    x = x.to(device)                       # тут можно просто ``.to("cuda")``
    z = x + y
    print(z)
    # с помощью``.to`` можно и изменить тип при перемещении
    print(z.to("cpu", torch.double))

tensor([[[1.1020, 1.8027, 1.4691, 1.1379, 1.3685],
         [1.9229, 1.5786, 1.5201, 1.4998, 1.8765],
         [1.4105, 1.2625, 1.5878, 1.7371, 1.0591],
         [1.0250, 1.8006, 1.3534, 1.9004, 1.9459],
         [1.1526, 1.1041, 1.0736, 1.1295, 1.4468]],

        [[1.3404, 1.7523, 1.7097, 1.7593, 1.5476],
         [1.1858, 1.2420, 1.8676, 1.7241, 1.0772],
         [1.4103, 1.6413, 1.5582, 1.4314, 1.2843],
         [1.8597, 1.3757, 1.9740, 1.7536, 1.2312],
         [1.5146, 1.4772, 1.6656, 1.2667, 1.9044]],

        [[1.8459, 1.5266, 1.1187, 1.4409, 1.5193],
         [1.9581, 1.7183, 1.6239, 1.0969, 1.2514],
         [1.4570, 1.8244, 1.0401, 1.0842, 1.0928],
         [1.3497, 1.3674, 1.1885, 1.4647, 1.3512],
         [1.7557, 1.9227, 1.3935, 1.6641, 1.1584]],

        [[1.2204, 1.9905, 1.2238, 1.6677, 1.2361],
         [1.6844, 1.6748, 1.2868, 1.7353, 1.6190],
         [1.5979, 1.6048, 1.3799, 1.0348, 1.6583],
         [1.1995, 1.9720, 1.2470, 1.0917, 1.1947],
         [1.6680, 1.5499,

<h3 style="text-align: center;">Autograd<b></b></h3>

Расшифровывается как Automatic Gradients (автоматическое взятие градиентов) - собственно, из названия понятно, что это модуль PyTorch, отвечающий за взятие производных.  

Возможно, для вас это бдет шок, но PyTorch (и любой фреймворк глубокого обучения) может продифференцировать функцию практически любой сложности.

Импортируем нужный класс:

In [140]:
from torch.autograd import Variable

Идея такая: оборачиваем тензор в класс Variable(), получаем тоже тензор, но он имеет способность вычислять себе градиенты.  

Если а - тензор, обёрнутый в Variable(), то при вызове a.backward() берутся градиенты по всем переменным, от которых зависит тензор a.

**ВНИМАНИЕ!**  

Если вы используете версию `pytorch 0.4.0` или более новую, то ***`torch.Tensor` и `torch.Variable` - одно и то же!*** То есть Вам больше не нужно оборачивать в `Variable()`, чтобы брать градиенты - они берутся и по `Tensor()` (`torch.Variable()` - deprecated).

Примеры:

In [141]:
x = torch.FloatTensor(3, 1).uniform_()
w = torch.FloatTensor(3, 3).uniform_() 
b = torch.FloatTensor(3, 1).uniform_()

x = Variable(x, requires_grad=True)
w = Variable(w, requires_grad=True)
b = Variable(b, requires_grad=False)

y = (w @ x).add_(b)

loss = y.sum()

# берём градиенты по всем "листьям" - в данном случае это тензоры x, w и b
loss.backward()

In [142]:
x.grad

tensor([[2.1250],
        [2.2211],
        [1.5210]])

In [143]:
w.grad

tensor([[0.9414, 0.0176, 0.4850],
        [0.9414, 0.0176, 0.4850],
        [0.9414, 0.0176, 0.4850]])

In [144]:
b.grad

In [145]:
y.grad

**Обратите внимание** - градиенты лежат в поле `.grad` у тех тензоров (Variable'ов), по которым брали эти градиенты. Градиенты **не лежат** в той Variable, от котороый они брались!

Получить тензор из `Variable()` можно с помощью поля `.data`:

In [146]:
x

tensor([[0.9414],
        [0.0176],
        [0.4850]], requires_grad=True)

In [147]:
x.data

tensor([[0.9414],
        [0.0176],
        [0.4850]])

### Задача 5

- Объявите тензор `a` размера (2, 3, 4) и тензор `b` размера (1, 8, 3), иницилизируйте их случайно равномерно (`.uniform_()`, как в примере выше)
- Создайте их копии на GPU, выведите их сумму и разность
- Затем измените форму тензора `b`, чтобы она совпадала с формой тензора `a`, получите тензор `c`  
- Переместите `c` на CPU, переместите `a` на CPU  
- Оберните их в `Variable()`
- Объявите тензор `L = torch.mean((c - a) `**` 2)` и посчитайте градиент `L` по `c` ( то есть $\frac{\partial{L}}{\partial{c}})$
- Получите тензор из `c` (то есть сейчас `c` - объект типа `Variable()`, вам нужно получить из него `FloatTensor()`)

In [14]:
a = torch.FloatTensor(2, 3, 4).uniform_().cpu()
b = torch.FloatTensor(1, 8, 3).uniform_().cpu()

a1 = a.cuda()
b1 = b.cuda()

print(f'A:{a}')
print(f'B:{b}')

c = b1.view(2, 3, 4).cpu()
print(f'C:{c}')

print(f'A+C:{a.add(c)}')
print(f'A-C:{a.sub(c)}')



L = torch.mean((c - a)**2)
print(f'L:{L}')


A:tensor([[[0.6510, 0.3127, 0.8571, 0.9410],
         [0.1830, 0.6180, 0.0317, 0.3616],
         [0.2873, 0.4691, 0.4750, 0.6957]],

        [[0.7929, 0.7905, 0.7799, 0.4884],
         [0.6295, 0.2740, 0.9761, 0.4105],
         [0.0260, 0.5226, 0.8031, 0.8702]]])
B:tensor([[[0.1208, 0.6384, 0.8591],
         [0.4767, 0.9610, 0.1262],
         [0.9367, 0.9183, 0.8538],
         [0.6971, 0.1358, 0.9308],
         [0.4036, 0.3156, 0.1632],
         [0.4267, 0.2726, 0.4272],
         [0.6717, 0.9652, 0.2184],
         [0.0807, 0.1066, 0.6171]]])
C:tensor([[[0.1208, 0.6384, 0.8591, 0.4767],
         [0.9610, 0.1262, 0.9367, 0.9183],
         [0.8538, 0.6971, 0.1358, 0.9308]],

        [[0.4036, 0.3156, 0.1632, 0.4267],
         [0.2726, 0.4272, 0.6717, 0.9652],
         [0.2184, 0.0807, 0.1066, 0.6171]]])
A+C:tensor([[[0.7718, 0.9511, 1.7163, 1.4177],
         [1.1441, 0.7442, 0.9683, 1.2799],
         [1.1411, 1.1661, 0.6109, 1.6265]],

        [[1.1965, 1.1062, 0.9430, 0.9151],
         [

<h3 style="text-align: center;">Полезные ссылки:<b></b></h3>

*1). Мегаполезная статья по PyTorch (на русском), на её основе делался этот ноутбук: https://habr.com/post/334380/*

*2). Туториалы от самих разработчиков фреймворка: https://pytorch.org/tutorials/beginner/blitz/tensor_tutorial.html#sphx-glr-beginner-blitz-tensor-tutorial-py*

*3). Статья на arXiv о сравнении фреймворков глубокого обучения: https://arxiv.org/pdf/1511.06435.pdf*

4). *Ещё туториалы: https://github.com/yunjey/pytorch-tutorial*

*5). Сайт Facebook AI Research - отдела, который разрабатывает PyTorch и другие крутые вещи в AI: https://facebook.ai/developers/tools*